pkg/kubelet/cri/remote/remote_runtime.go

```diff
@@ -1208,3 +1209,33 @@ func (r *remoteRuntimeService) CheckpointContainer(options *runtimeapi.Checkpoin
 
 	return nil
 }
+
+func (r *remoteRuntimeService) GetContainerEvents(containerEventsCh chan *runtimeapi.ContainerEventResponse) error {
+
+	if !r.useV1API() {
+		return nil
+	}
+
+	containerEventsStreamingClient, err := r.runtimeClient.GetContainerEvents(context.Background(), &runtimeapi.GetEventsRequest{})
```

Comment:
what happens when this times out?  isnt this subject to max time of open grpc connection?

Reply:
Added code to retry on error and force a relisting to update the cache and (pod and container running) metric, https://github.com/kubernetes/kubernetes/pull/111384/files#diff-465aae123f6877816ff0428968fc74f503fe15b27b434480bf57b79e62c7fa2fR118

Reply:
This is an area where being able to trigger this in a test (ideally above unit) will expose low frequency errors sooner.

Does the event stream include heartbeats?  There are lots of failure modes for event streams that involve not explicitly getting events even though the connection is still open.

If we detect an anomaly during relist (missed event), we may want to force this connection closed and reopen it.

Reply:
Maybe I should have asked this before - but how are you synchronizing this with the "List" action so that you don't lose events between the time of List and when you invoke GetContainerEvents?  You need an equivalent to the `ResourceVersion` construct and the same guarantees watch provides to do that without gaps - and I don't see that here or in the enhancement?

Reply:
> If we detect an anomaly during relist (missed event), we may want to force this connection closed and reopen it.

This is an interesting idea.

Reply:
> Maybe I should have asked this before - but how are you synchronizing this with the "List" action so that you don't lose events between the time of List and when you invoke GetContainerEvents? You need an equivalent to the `ResourceVersion` construct and the same guarantees watch provides to do that without gaps - and I don't see that here or in the enhancement?

We are getting the grpc streaming client at the kubelet start up and then keep receiving the events using `containerEventsStreamingClient.Recv()` call below

Reply:
> Does the event stream include heartbeats? There are lots of failure modes for event streams that involve not explicitly getting events even though the connection is still open.
> 

We are simply using grpc streaming client, and depend on their mechanism to maintain and verify a healthy stream. We do get error back from that client should they detect anything wrong.